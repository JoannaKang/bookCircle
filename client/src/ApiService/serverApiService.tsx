import { User, ActivityLog, Book } from '../Interfaces'

const URL = 'http://localhost:3001/'

async function getUser(name: string): Promise<User> {
  const user = await fetch(URL+'getUser/'+name)
    .then(response => response.json())

  return user[0];
}

async function getFriendName(name: string): Promise<Object> {
  const friendsName = await fetch(URL+'getFriendsNames/'+name)
    .then(response => response.json())
  return friendsName.name;
}

async function acceptFriend(activity: ActivityLog): Promise<Object> {
  const result = await fetch(URL+'confirmFriend', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(activity)
  })
  .then(response => response.json())
  return result;
}

async function rejectFriend(activity: ActivityLog): Promise<Object> {
  const result = await fetch(URL+'rejectFriendRequest', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(activity)
  })
  .then(response => response.json())
  return result;
}

type BookRequest = {
  user: User,
  book: Book,
  friendId: any
}

async function requestBook(obj: BookRequest): Promise<User> {
  const result = await fetch(URL+'requestBook', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(obj)
  })
  .then(response => response.json())
  return result;
}

async function acceptBookRequest(activity: ActivityLog): Promise<Object> {
  const result = await fetch(URL+'acceptBookRequest', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(activity)
  })
  .then(response => response.json())
  return result;
}

async function rejectBookRequest(activity: ActivityLog): Promise<Object> {
  const result = await fetch(URL+'rejectBookRequest', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(activity)
  })
  .then(response => response.json())
  return result;
}

async function deleteMessage(activity: ActivityLog): Promise<Object> {
  const result = await fetch(URL+'removeActivityLogElement', {
    method: 'DELETE',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(activity)
  })
  .then(response => response.json())
  return result;
}

export {
  getUser,
  getFriendName,
  acceptFriend,
  rejectFriend,
  requestBook,
  acceptBookRequest,
  rejectBookRequest,
  deleteMessage
}