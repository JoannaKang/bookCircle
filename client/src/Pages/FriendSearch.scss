.searchHeader {
  font-weight: 300;
  margin: 1rem 2.2rem;
}

h2 {
  margin-left: 40px;
}