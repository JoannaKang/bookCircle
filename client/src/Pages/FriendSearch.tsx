import React, { FunctionComponent, useState } from "react"
import FriendsSearchResults from '../Components/FriendsSearchResults'
import { searchFriend } from '../ApiService/serverApiService'
import {User} from '../Interfaces'

type FriendSearchProps = {
  handleAddFriend: Function,
  user: User
}

const FriendSearch: FunctionComponent<FriendSearchProps> = ({handleAddFriend, user}) => {
  const [searchName, setSearchName] = useState('')
  const [searchResults, setSearchResults] = useState<any>()
  const [isSearch, setIsSearch] = useState(false);


  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const result: any = await searchFriend(searchName);
    setSearchName('');
    setSearchResults(result)
    setIsSearch(true)
  };

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setSearchName(e.target.value)
  };


  return (
    <div>
      <h1 className='searchHeader'>Find Friends:</h1>
      <form className='searchForm' action="submit" onSubmit={handleSubmit}>
        <div className='inputAndBtnContainer'>
          <input onChange={handleChange} className='searchInput' type="text" value={searchName} placeholder={`search by username...`}/>
          <button className="searchBtn">search</button>
        </div>
      </form>
    { isSearch && <FriendsSearchResults users={searchResults} handleAddFriend={handleAddFriend} user={user} />
  }
    </div>
  )
}

export default FriendSearch;