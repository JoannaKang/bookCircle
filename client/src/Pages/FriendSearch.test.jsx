import FriendSearch from './FriendSearch'
import { render, fireEvent, getByTestId } from '@testing-library/react'
import { searchFriend } from '../ApiService/serverApiService'

describe('FriendSearch', () => {

    const user = {
        "books": [],
        "friends": [],
        "pendingFriends": [],
        "name": "Sooyeon",
        "password": "abc",
        "yearlyTarget": 7,
        "activityLog": [],
    }

    it('matches the snapshot when user is provided', () => {
        const { container } = render(<FriendSearch user={user}/>)
        expect(container).toMatchSnapshot()
    })

    it('Matches Snapshot when search button is Clicked and searchFriend Function is triggered', () => {
        const searchFriend = jest.fn();
        const handleAddFriend = jest.fn();
        const { getByTestId } = render(<FriendSearch user={user} handleAddFriend={handleAddFriend}/>);
        fireEvent.click(getByTestId("searchClick"));
        const { container } = render(<FriendSearch user={user} handleAddFriend={handleAddFriend}/>);
        expect(container).toMatchSnapshot();
        expect(searchFriend);
    })

})