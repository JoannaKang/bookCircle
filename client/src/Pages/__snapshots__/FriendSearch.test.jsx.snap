// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FriendSearch Matches Snapshot when search button is Clicked and searchFriend Function is triggered 1`] = `
<div>
  <div>
    <h1
      class="searchHeader"
    >
      Find Friends:
    </h1>
    <form
      action="submit"
      class="searchForm"
    >
      <div
        class="inputAndBtnContainer"
      >
        <input
          class="searchInput"
          placeholder="search by username..."
          type="text"
          value=""
        />
        <button
          class="searchBtn"
          data-testid="searchClick"
        >
          search
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`FriendSearch matches the snapshot when user is provided 1`] = `
<div>
  <div>
    <h1
      class="searchHeader"
    >
      Find Friends:
    </h1>
    <form
      action="submit"
      class="searchForm"
    >
      <div
        class="inputAndBtnContainer"
      >
        <input
          class="searchInput"
          placeholder="search by username..."
          type="text"
          value=""
        />
        <button
          class="searchBtn"
          data-testid="searchClick"
        >
          search
        </button>
      </div>
    </form>
  </div>
</div>
`;
